### PR TITLE
Review Fixes and Enhancements to the SDK

### PR DESCRIPTION
**Pull request by karthikShiva1997**
1) setRange and getRange methods added to PriceBookPricing

2) Fixed an issue where the right error was not thrown from the uploadPhoto method when file size 
     exceeds 20mb

3) Added methods to extract the error message from its enclosing enum case

4) JSON root Key now has a default value of DATA since most handlers use this value . Other handlers 
     can override this value using setJSONRootKey method

5) Response Root Key is now Non-Optional in both CommonAPIResponse and APIResponse since it 
     has a  default value of DATA

6) Fixed an issue where the  initialiseRequest method of the API Request Class would stall when   
     isOauth == false 

7) ZCRMSubForm setters now have optional parameters to accomodate nil values

8) Added validation to prevent uploading files other than the supported ones (GIF,PNG,BMP,JPG)  in CRM

### CODE REVIEW

File: ZCRMiOS/Classes/APIResponse.swift [hunk 1]
```diff
@@ -111,7 +114,7 @@ public class FileAPIResponse : APIResponse
     init(response: HTTPURLResponse, tempLocalUrl: URL?) throws
     {
         self.tempLocalUrl = tempLocalUrl!
-        try super.init(response: response, responseData: nil, responseJSONRootKey : nil )
+        try super.init(response: response,responseData: nil,responseJSONRootKey:"noRootKey")
```
Comment (vijayakrishnap) [inline ZCRMiOS/Classes/APIResponse.swift]: Why is this some string instead of nil?

File: ZCRMiOS/Classes/ZCRMSubformRecord.swift [hunk 2]
```diff
@@ -39,7 +39,7 @@ public class ZCRMSubformRecord : ZCRMEntity
 		return self.owner
 	}
 	
-	internal func setModifiedTime( modifiedTime : String )
+	internal func setModifiedTime( modifiedTime : String? )
```
Comment (vijayakrishnap) [inline ZCRMiOS/Classes/ZCRMSubformRecord.swift]: Why is this changed to optional? In what case can this come as null ?

**Review by vijayakrishnap — CHANGES_REQUESTED** [review]

File: ZCRMiOS/Classes/APIResponse.swift (hunk 1 above)
Comment (karthikShiva1997) [reply ZCRMiOS/Classes/APIResponse.swift]: This because the super Class (CommonAPIResponse) is of non - Optional type for the property responseJSONRootKey. It was made non-optional because a non-optional value of DATA is always passed to the APIResponse Class by the Handler classes since DATA is the default JSON Root key for all Handlers unless they over write the value

Comment (vijayakrishnap) [reply ZCRMiOS/Classes/APIResponse.swift]: Then lets move this to some constant file and have the dummy value as "file" instead of "noRootKey" so that in future if we have more such types, we can have keys for each type.

Comment (karthikShiva1997) [reply ZCRMiOS/Classes/APIResponse.swift]: Okay Anna ! The changes will be reflected in the next commit

**Comment by karthikShiva1997** [comment]
1. Reverted Back ZCRMSubForm setters to have non-optional setters .
2. JSON Root Keys now have their own Struct in Common Utils Class

**Comment by vijayakrishnap** [comment]
@karthikShiva1997  Resolve merge conflicts.

**Review by vijayakrishnap — APPROVED** [review]
